File: 레거시/BOJ/Kotlin/30000/30024.kt
import java.util.*

val dx = intArrayOf(0, 0, 1, -1)
val dy = intArrayOf(1, -1, 0, 0)

fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readln().split(" ").map { it.toInt() }
    val value = Array(n) { IntArray(m) { 0 } }
    repeat(n) { i -> value[i] = readln().split(" ").map { it.toInt() }.toIntArray() }
    val k = readln().toInt()

    val pq = PriorityQueue<Triple<Int, Int, Int>>(compareByDescending { it.first })  // (value, x, y)
    val visited = Array(n) { BooleanArray(m) { false } }

    for (i in 0 until n) {
        for (j in 0 until m) {
            if (i == 0 || j == 0 || i == n - 1 || j == m - 1) {
                pq.add(Triple(value[i][j], i, j))  // insert outside value
                visited[i][j] = true
            }
        }
    }

    for (i in 0 until k) {
        val (_, x, y) = pq.poll()
        println("${x + 1} ${y + 1}")

        for (j in 0..3) {
            val nx = x + dx[j]
            val ny = y + dy[j]

            if (nx < 0 || nx >= n || ny < 0 || ny >= m || visited[nx][ny]) continue
            pq.add(Triple(value[nx][ny], nx, ny))
            visited[nx][ny] = true
        }
    }
}

File: 레거시/BOJ/C++/30000/30024.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

const int dx[] = {0, 0, 1, -1};
const int dy[] = {1, -1, 0, 0};

int main() {
    int n, m;
    cin >> n >> m;
    vector<vector<int>> value(n, vector<int>(m, 0));

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cin >> value[i][j];
        }
    }

    int k;
    cin >> k;

    priority_queue<pair<int, pair<int, int>>> pq; // (value, (x, y))
    vector<vector<bool>> visited(n, vector<bool>(m, false));

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (i == 0 || j == 0 || i == n - 1 || j == m - 1) {
                pq.push({value[i][j], {i, j}}); // insert outside value
                visited[i][j] = true;
            }
        }
    }

    for (int i = 0; i < k; i++) {
        int x, y, val;
        val = pq.top().first;
        x = pq.top().second.first;
        y = pq.top().second.second;
        pq.pop();
        cout << x + 1 << " " << y + 1 << "\n";

        for (int j = 0; j < 4; j++) {
            int nx = x + dx[j];
            int ny = y + dy[j];

            if (nx < 0 || nx >= n || ny < 0 || ny >= m || visited[nx][ny]) continue;
            pq.push({value[nx][ny], {nx, ny}});
            visited[nx][ny] = true;
        }
    }

    return 0;
}